## Publicly-accessible AIT-formatted taxonomies

We have converted several published brain cell type taxonomies to AIT format to enable community mapping (e.g., with [scrattch.mapping](https://github.com/AllenInstitute/scrattch.mapping)). The initial set of taxonomies are from the [Cell Types Database: RNA-Seq Data](https://portal.brain-map.org/atlases-and-data/rnaseq) on Allen Brain Map and related Allen Institute papers and data sets.  Future updates may include additional cell type taxonomies from other BICAN groups or other community sources, both in the brain and across other organ systems.  

Cell type taxonomies are listed roughly from oldest to newest by data of **data set** publication. All listed taxonomies follow the [AIT schema](https://github.com/AllenInstitute/AllenInstituteTaxonomy/tree/main/schema) (version listed in-line) and should be compatible with the [scrattch packages](https://github.com/AllenInstitute/scrattch) for cell type analysis and mapping. All taxonomies also (1) have low quality cells and/or clusters omitted, and are set up for correlation and Seurat mapping, but are NOT set up for tree mapping (see [scrattch.mapping](https://github.com/AllenInstitute/scrattch.mapping) for details).  Some are set up for MapMyCells mapping and some are not (see additional notes).  Additional notes specific to subsets of taxonomies are listed below the table, and linked to letters next to the file names in the **AIT file(s)** column.

*Please [submit an issue](https://github.com/AllenInstitute/AllenInstituteTaxonomy/issues) or post a topic to the [Allen Brain Map Community Forum](https://community.brain-map.org/) if you run into any issues with any or these files or if you'd like to request other cell type taxonomies in AIT format.*
<br>
<br>

### Table of available taxonomies

| Data Set Title | Species | Original Location | Publication(s) | AIT file(s) | Supporting file(s)|
|:-----|:------|:------|:------|:----|:----|
| **V1 & ALM - SMART-seq (2018)**  | Mouse |  [Allen Brain Map](https://portal.brain-map.org/atlases-and-data/rnaseq/mouse-v1-and-alm-smart-seq)    |  [Tasic, et al, 2018](https://www.nature.com/articles/s41586-018-0654-5) | **[AIT v1.1.2 (4 Apr 2025)](https://allen-cell-type-taxonomies.s3.us-west-2.amazonaws.com/Mouse_VISp_ALM_SMART_seq_04042025.h5ad)** (a,c,e,f)  | [Conversion script](https://raw.githubusercontent.com/AllenInstitute/AllenInstituteTaxonomy/main/conversion_scripts/AIT_script_Mouse_VISp_ALM_SMART_seq_04042025.r);<br> [Dendrogram](https://github.com/AllenInstitute/AllenInstituteTaxonomy/raw/refs/heads/main/conversion_data/Mouse_VISp_ALM_SMART_seq_04042025_dend.RData);<br> [TSNE coordinates](https://github.com/AllenInstitute/AllenInstituteTaxonomy/raw/refs/heads/main/conversion_data/Mouse_VISp_ALM_SMART_seq_04042025_tsne.feather)  |
| **MTG - SMART-seq (2018)**   | Human |  [Allen Brain Map](https://portal.brain-map.org/atlases-and-data/rnaseq/human-mtg-smart-seq)  |  [Hodge, Bakken, et al, 2019](https://www.nature.com/articles/s41586-019-1506-7) | ~~**[AIT v1.1.2 (4 Apr 2025)](https://allen-cell-type-taxonomies.s3.us-west-2.amazonaws.com/Human_MTG_SMART_seq_04042025.h5ad)** (a,e,f,g)~~; **[AIT v1.1.2 (8 Aug 2025)](https://allen-cell-type-taxonomies.s3.us-west-2.amazonaws.com/Human_MTG_SMART_seq_08082025.h5ad)** (a,e,f,g,j)| [Conversion script](https://raw.githubusercontent.com/AllenInstitute/AllenInstituteTaxonomy/main/conversion_scripts/AIT_script_Human_MTG_SMART_seq_04042025.r);<br> [Dendrogram](https://github.com/AllenInstitute/AllenInstituteTaxonomy/raw/refs/heads/main/conversion_data/Human_MTG_SMART_seq_04042025_dend.RData);<br> [TSNE coordinates](https://github.com/AllenInstitute/AllenInstituteTaxonomy/raw/refs/heads/main/conversion_data/Human_MTG_SMART_seq_04042025_tsne.csv);<br> [Cluster information](https://github.com/AllenInstitute/AllenInstituteTaxonomy/raw/refs/heads/main/conversion_data/Human_MTG_SMART_seq_04042025_cluster_info.csv);<br> [Cell by **isoform** matrix](https://allen-cell-type-taxonomies.s3.us-west-2.amazonaws.com/Hodge2019_isoforms.h5ad)  |
| **Comparative LGN (2018)**  | Human, mouse, macaque |  [Allen Brain Map](https://portal.brain-map.org/atlases-and-data/rnaseq/comparative-lgn)    |  [Bakken, van Velthoven, Menon, et al, 2021](https://elifesciences.org/articles/64875) | **[Human AIT v1.1.2 (4 Apr 2025)](https://allen-cell-type-taxonomies.s3.us-west-2.amazonaws.com/Human_LGN_SMART_seq_04042025.h5ad)** (d); <br> **[Mouse AIT v1.1.2 (4 Apr 2025)](https://allen-cell-type-taxonomies.s3.us-west-2.amazonaws.com/Mouse_LGN_SMART_seq_04042025.h5ad)** (d); <br> **[Macaque AIT v1.1.2 (4 Apr 2025)](https://allen-cell-type-taxonomies.s3.us-west-2.amazonaws.com/Macaque_LGN_SMART_seq_04042025.h5ad)** (d)  | [Human conversion script](https://raw.githubusercontent.com/AllenInstitute/AllenInstituteTaxonomy/main/conversion_scripts/AIT_script_Human_LGN_SMART_seq_04042025.r);<br> [Mouse conversion script](https://raw.githubusercontent.com/AllenInstitute/AllenInstituteTaxonomy/main/conversion_scripts/AIT_script_Mouse_LGN_SMART_seq_04042025.r);<br> [Macaque conversion script](https://raw.githubusercontent.com/AllenInstitute/AllenInstituteTaxonomy/main/conversion_scripts/AIT_script_Macaque_LGN_SMART_seq_04042025.r)   |
| **Multiple Cortical Areas - SMART-seq (2019)**  | Human |  [Allen Brain Map](https://portal.brain-map.org/atlases-and-data/rnaseq/human-multiple-cortical-areas-smart-seq)  |  [Jorstad, et al, 2023](https://www.science.org/doi/10.1126/science.adf6812) | **[AIT v1.1.2 (4 Apr 2025)](https://allen-cell-type-taxonomies.s3.us-west-2.amazonaws.com/Human_neocortex_SMART_seq_04042025.h5ad)** (a,e,i)  | [Conversion script](https://raw.githubusercontent.com/AllenInstitute/AllenInstituteTaxonomy/main/conversion_scripts/AIT_script_neocortex_SMART_seq_04042025.r);<br> [Dendrogram](https://github.com/AllenInstitute/AllenInstituteTaxonomy/raw/refs/heads/main/conversion_data/Human_neocortex_SMART_seq_04042025_dend.RData) |
| **M1 - 10x genomics (2020)**  | Human |  [Allen Brain Map](https://portal.brain-map.org/atlases-and-data/rnaseq/human-m1-10x)    |  [Bakken, et al, 2021](https://www.nature.com/articles/s41586-021-03465-8) | **[AIT v1.1.2 (4 Apr 2025)](https://allen-cell-type-taxonomies.s3.us-west-2.amazonaws.com/Human_M1_10X_seq_04042025.h5ad)** (a)  | [Conversion script](https://raw.githubusercontent.com/AllenInstitute/AllenInstituteTaxonomy/main/conversion_scripts/AIT_script_Human_M1_10X_seq_04042025.r);<br> [Cell Type Knowledge Explorer](https://knowledge.brain-map.org/celltypes/CCN201912131)  |
| **Whole Cortex and Hippocampus (2021)**  | Mouse |  Allen Brain Map<br>([10X genomics](https://portal.brain-map.org/atlases-and-data/rnaseq/mouse-whole-cortex-and-hippocampus-10x); [SMART-seq](https://portal.brain-map.org/atlases-and-data/rnaseq/mouse-whole-cortex-and-hippocampus-smart-seq))    |  [Yao, et al, 2021](https://www.sciencedirect.com/science/article/pii/S0092867421005018) | **[10X genomics AIT v1.1.2 (4 Apr 2025)](https://allen-cell-type-taxonomies.s3.us-west-2.amazonaws.com/Mouse_cortex_hippocampus_10X_seq_04042025.h5ad)** (c,h) <br> **[SMART-seq AIT v1.1.2 (4 Apr 2025)](https://allen-cell-type-taxonomies.s3.us-west-2.amazonaws.com/Mouse_cortex_hippocampus_SMART_seq_04042025.h5ad)** (b,c)  | [10x genomics conversion script](https://raw.githubusercontent.com/AllenInstitute/AllenInstituteTaxonomy/main/conversion_scripts/AIT_script_Mouse_cortex_hippocampus_10X_seq_04042025.r);<br> [SMART-seq conversion script](https://raw.githubusercontent.com/AllenInstitute/AllenInstituteTaxonomy/main/conversion_scripts/AIT_script_Mouse_cortex_hippocampus_SMART_seq_04042025.r)   |
| **MTG - 10x SEA-AD (2022)**  | Human |  [Allen Brain Map](https://portal.brain-map.org/atlases-and-data/rnaseq/human-mtg-10x_sea-ad)  |  [Gabitto, Travaglini, et al, 2024](https://www.nature.com/articles/s41593-024-01774-5) | **[AIT v1.1.2 (4 Apr 2025)](https://allen-cell-type-taxonomies.s3.us-west-2.amazonaws.com/Human_MTG_SEAAD_04042025.h5ad)** (a)  | [Conversion script](https://raw.githubusercontent.com/AllenInstitute/AllenInstituteTaxonomy/main/conversion_scripts/AIT_script_Human_MTG_SEAAD_04042025.r)  |

#### Additional taxonomy-specific notes

* **(a)**: Taxonomy set up for MapMyCells (also called "hierarchical" and "cell_type_mapper") mapping set up using all levels<br>
* **(b)**: Taxonomy set up for MapMyCells (also called "hierarchical" and "cell_type_mapper") mapping set up using all levels EXCEPT neighborhood<br>
* **(c)**: Original dendrograms were not binary trees and therefore are incompatible with scrattch v1.1.2; therefore, NEW dendrograms are created for these AIT files that do NOT match the original<br>
* **(d)**: Original dendrograms were not available; therefore, NEW dendrogram are created for these AIT files that does NOT match the original<br>
* **(e)**: Requires download of files only found on this repo<br>
* **(f)**: Cluster names match what is in the paper, which is different from what is on Allen Brain Map for a few clusters<br>
* **(g)**: Subclasses align with reported "pCL name"<br>
* **(h)**: Taxonomy is subsampled to 1000 cells per cluster so count matrices can be loaded into R<br>
* **(i)**: Taxonomy corresponds to Allen Brain Map data and exploration tool, which is an **OLD VERSION** of the taxonomy presented in Jorstad et al 2023<br>
* **(j)**: Taxonomy updated 8/8/25 to correct error in TSNE coordinates. 70 Ensembl IDs also have changed, leading to minor changes in MapMyCells stats and markers. Everything else in the the file is unchanged.

